### DISCUSSION TITLE
Optimizing model for precision and recall

### DISCUSSION THREAD

**Discussion by cw-18** [microsoft/FLAML#953] [Q&A]
Can automl optimize models for precision and recall? It doesn't seem like precision and recall are available options for metrics. 

Ex: 
automl_settings = {
    "time_budget": 30, 
    "metric": 'recall',  
    "task": 'classification',
    "log_file_name": "cell.log",
}

gives the error 
'recall is not an built-in sklearn metric and nlp is not installed. Currently built-in sklearn metrics are: r2, rmse, mae, mse, accuracy, roc_auc, roc_auc_ovr, roc_auc_ovo,log_loss, mape, f1, micro_f1, macro_f1, ap. If the metric is an nlp metric, please pip install flaml[nlp] ', 'or pass a customized metric function to AutoML.fit(metric=func)'

Based on the error message, it doesn't seem like automl includes 'precision' and 'recall' as available metrics. 

Based on https://learn.microsoft.com/en-us/azure/machine-learning/how-to-understand-automated-ml, I would think that 'precision' and 'recall' would be included. 

In addition, it's interesting because metrics like f1, which rely on precision and recall, are included as available metrics so I wonder why precision and recall wouldn't be included as available metrics. 
I guess in practice, it's favorable to optimize for both precision and recall (hence optimize for metrics like f1), over optimize for solely precision or solely recall.

Am I missing something?

**Comment by EgorKraevTransferwise** [comment]
Optimizing just one of them leads to degenerate solutions: using the definitions from [here](https://developers.google.com/machine-learning/crash-course/classification/precision-and-recall) just classifying everything as a positive gives a precision of 1.0 since there are no false negatives; likewise, classifying everything as negative gives a recall of 1.0 as there are no false positives. 
So it makes sense to either trade off between them, eg with some f1 flavor, or use a different metric, eg one of the ones you list or a custom one.
